lib/drivers.js

```diff
@@ -0,0 +1,198 @@
+import path from 'path';
+import _ from 'lodash';
+import log from './logger';
+import { fs } from 'appium-support';
+import { existsSync } from 'fs';
+
+// TODO the location of the driver data file should be configurable
+const DRIVERS_JSON = path.resolve(__dirname, '..', '..', '.drivers.json');
+
+// On Appium launch, figure out which drivers are installed
+let INSTALLED_DRIVERS = {};
+
+// This is a map of driver names to npm packages representing those drivers.
+// The drivers in this list will be available to the CLI so users can just
+// type 'appium driver install 'name'', rather than having to specify the full
+// npm package. I.e., these are the officially recognized drivers.
+const KNOWN_DRIVERS = {
+  uiautomator2: 'appium-uiautomator2-driver',
+  xcuitest: 'appium-xcuitest-driver',
+  youiengine: 'appium-youiengine-driver',
+  windows: 'appium-windows-driver',
+  mac: 'appium-mac-driver',
+  espresso: 'appium-espresso-driver',
+  tizen: 'appium-tizen-driver',
+};
+
+const INSTALL_TYPE_NPM = 'npm';
+const INSTALL_TYPE_LOCAL = 'local';
+const INSTALL_TYPE_GITHUB = 'github';
+const INSTALL_TYPE_GIT = 'git';
+const INSTALL_TYPES = [
+  INSTALL_TYPE_GIT,
+  INSTALL_TYPE_GITHUB,
+  INSTALL_TYPE_LOCAL,
+  INSTALL_TYPE_NPM
+];
+
+function validateDriversJson () {
+  try {
+    INSTALLED_DRIVERS = require(DRIVERS_JSON);
+  } catch (err) {
+    if (existsSync(DRIVERS_JSON)) {
+      throw new Error(`Appium had trouble loading the driver installation ` +
+                      `cache file (${DRIVERS_JSON}). Ensure it exists and is ` +
+                      `readable. Specific error: ${err.message}`);
+    }
+
+    // if the file just doesn't exist, call it empty
+    return {};
+  }
+
+  const automationNames = [];
+  let msg = '';
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const platformNames = INSTALLED_DRIVERS[driverName].platformNames;
+    if (!_.isArray(platformNames)) {
+      msg += `Entry for '${driverName}' has missing or ` +
```

Comment:
I would also print the actual values in the error message

Reply:
going to refactor this whole bit so it's cleaner and can allow validation to proceed when some drivers are valid and some aren't.